File: db/connection.js
const mysql = require('mysql2/promise');
require('dotenv').config();

const pool = mysql.createPool({
    host:               process.env.DB_HOST     || 'localhost',
    port:               parseInt(process.env.DB_PORT || '3306'),
    user:               process.env.DB_USER     || 'root',
    password:           process.env.DB_PASSWORD || 'yuan@123',
    database:           process.env.DB_NAME     || 'threatpulse',
    waitForConnections: true,
    connectionLimit:    10,
    queueLimit:         0,
    charset:            'utf8mb4',
    timezone:           'Z',          // store/return UTC
});

// Verify connectivity on startup
pool.getConnection()
    .then(conn => {
        console.log('[DB] Connected to MySQL — threatpulse database');
        conn.release();
    })
    .catch(err => {
        console.error('[DB] Connection failed:', err.message);
        process.exit(1);
    });

module.exports = pool;

File: db/connection.py
"""
PyMySQL connection pool for ThreatPulse.
All DB queries use get_db() as a FastAPI dependency.
"""
import os
import pymysql
import pymysql.cursors
from dotenv import load_dotenv
from contextlib import contextmanager

load_dotenv()

DB_CONFIG = {
    "host":    os.getenv("DB_HOST", "localhost"),
    "port":    int(os.getenv("DB_PORT", 3306)),
    "user":    os.getenv("DB_USER", "root"),
    "password": os.getenv("DB_PASSWORD", "yuan@123"),
    "database": os.getenv("DB_NAME", "threatpulse"),
    "charset": "utf8mb4",
    "cursorclass": pymysql.cursors.DictCursor,
    "autocommit": True,
}


def get_connection() -> pymysql.connections.Connection:
    """Open a new PyMySQL connection."""
    return pymysql.connect(**DB_CONFIG)


@contextmanager
def get_db():
    """Context manager: yields a connection, closes on exit."""
    conn = get_connection()
    try:
        yield conn
    finally:
        conn.close()


def get_db_dep():
    """FastAPI dependency that yields a DB connection per request."""
    with get_db() as conn:
        yield conn
